SDK Authentication

hernantain opened the pull request:
**What**: Implementing `TruStar` and `ApiClient` main methods for the SDK to be able to authenticate with TruSTARs API. TruStar is the main class that will forward all requests to ApiClient (maybe directly or maybe using a manager to handle logic with models. Managers are not implemented yet. Check `TruStar` constructor comments).

**Why**: These classes/methods are needed to bootstrap the SDK and be able to add more methods in the future. Our next objective is to use the JS generated files on ServiceNow.

**How**: Hitting `oauth/token` endpoint to retrieve token and using that token on header for future requests. At the moment the only other endpoint being hit is the `ping` endpoint to test that the SDK is working ok. 

Jira: https://trustar.atlassian.net/browse/EN-5145

To be able to test it, in your terminal:
 * Clone repo and checkout `feat/authentication-methods/EN-5145` branch.
 * Run `yarn install` in project's root.
 * Generate a `trustar.config.json` file containing

```json
{
  "user_api_key": <API_KEY>,
  "user_api_secret": <API_SECRET>
}
```
 * Run `tsc`
 * Run `node trustar/trusar.js`

You should see printed on the console a `pong`.

marcos-vivar commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+  ];
+  auth_endpoint: string;
+  api_endpoint: string;
+  user_api_key: string;
```
Hi @hernantain!
I would like to point out that we need to add `the station_base_url` for future implementation of report's deeplinking.

hernantain replied:
Will add it!

marcos-vivar commented on trustar/trustar.ts:
```diff
@@ -0,0 +1,139 @@
+   * @param path path including filename to json file
+   */
+  readAndGetJSON(path: string): { [key: string]: any } {
+    let jsonString: string = fs.readFileSync(path, "utf-8");
```
Question: Are we able to do something like:
```
try {
   let jsonString: string = fs.readFileSync(path, "utf-8");
   return JSON.parse(jsonString);
  } catch(e) {
   return {};
  }
```

hernantain replied:
It's possible but I think the exception should be handled on the caller method. In this case the `readAndGetJSON` method is being called by `readConfigFile` and `readVersionFile` methods and it's trying to read 2 files that should be there. And I think the error it's descriptive enough

marcos-vivar commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+   * @param response HTTP Response
+   */
+  getErrorMessage(status: number, response: request.Response): string {
+    let trace_id: string = response.header["trace-id"]!;
```
Can this optional's unwrapping be done with a try/catch syntax?

marcos-vivar commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+    method: string,
+    path: string,
+    headers?: object,
+    params?: object,
```
Question: Can we default our `params` to something like:  `params? = {}: object`?

hernantain replied:
Yes, we can define `params: object = {}`  without the `?`.  Because typescript doesn't let you specify a default value and the `?` token that means undefined OR object (in this case). I will change it

marcos-vivar commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+   * or auth_endpoint to interact with TruSTAR API.
+   */
+  constructor(config: TrustarConfig) {
+    this.auth_endpoint = config["auth_endpoint"]!;
+    this.api_endpoint = config["api_endpoint"]!;
+    this.user_api_key = config["user_api_key"];
+    this.user_api_secret = config["user_api_secret"];
+    this.client_type = config["client_type"]!;
+    this.verify = config["verify"]!;
+    this.retry = config["retry"]!;
+    this.max_wait_time = config["max_wait_time"]!;
+    this.client_metatag = config["client_metatag"];
+    this.client_version = config["client_version"]!;
```
Should we declare all these strings as global constants and define them once?

jagrvargen commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+    while (!attempted || !retry) {
+      let is_json: boolean = ["POST", "PUT"].includes(method);
+      let base_headers: object = await this.getHeaders(is_json);
+      if (headers) {
```
We may have to be careful with conditions like this. JS is more loose than Python, so we should probably be more explicit with our conditions. For example, I ran this code in a REPL:
```
let x = {};
if (x) {
  console.log('works')
}
```
And it printed works to the console.

hernantain replied:
yes, good catch, in this case it will not break anything as it is doing an "union" of both objects (in case of repeated properties, the ones in `headers` will prevail). But will try to see for a safer way to do it, just in case

jagrvargen commented on trustar/trustar.ts:
```diff
@@ -0,0 +1,139 @@
+    return await this.apiClient.get("ping");
+  }
+
+  // WILL DELETE THIS
```
Do you still need to delete this or are we keeping it for now?

hernantain replied:
Yup, I left it just in case anyone wanted to test the code. I'm gonna remove it before merging

hernantain commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+    let retry: boolean = this.retry;
+    let attempted: boolean = false;
+    let response: request.Response;
+    while (!attempted || !retry) {
```
Yes! Good catch

hernantain commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+   *
+   * @returns The headers object.
+   */
+  async getHeaders(is_json: boolean = false): Promise<object> {
```
yeah, I've been a little inconsistent with variable naming styles, I'm going to change them all to camelCase

hernantain commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+  client_metatag: string;
+  client_version: string;
+  token: string | null;
+
```
In the mean_time I'd leave it. I haven't seen an easy way for super agent to handle proxies as it is done with the `requests` library on the python SDK. Maybe there is a way or maybe we should handle that on a different way? Same for the `verify` parameter to do a SSL verification.

hernantain replied:
I removed http_proxy, https_proxy and verify variables for now as they are not being used. We can add them later if needed.

leomestizo commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+
+      this.token = res.body["access_token"];
+    } catch (err) {
+      // console.debug(err.response["error"]);
```
:robot: Commented line :grin:

leomestizo commented on trustar/api_client.ts:
```diff
@@ -0,0 +1,232 @@
+          .query(request_params)
+          .send(data);
+
+        attempted = true;
```
Question: Will the `while` stop if there's an error in the request? Maybe you have to also add this line in the `catch`? Or even better to put it a `finally` clause.

hernantain replied:
@leomestizo It depends.

For what I've seen superagent raises an error if the http status code is different from 200 or even if the response is not JSON formatted. 

That's why I do multiple checks to see if there was an error related to JSON, expired token, 429s. In those cases the error will be handled accordingly., 

If effectively was a Network Error, or similar, it will stop execution and raise the Error.

leomestizo commented:
It looks good to me man! I added a couple of minor things/questions, but looks great overall.

hernantain commented:
@leomestizo @jagrvargen @marcos-vivar Thanks for reviewing folks, I addressed most of your comments, so feel free to take another look when you have some free time.

jagrvargen reviewed (dismissed):
lgtm -- let's get started on the other stuff

marcos-vivar reviewed (dismissed):
Outstanding work, @hernantain! 👍🏽 

Is it possible to add something trivial, to start attacking test cases, for `tests/trustar.test.ts`? Maybe we can use your own lines, as:
```
let ts = new TruStar();
ts.print();
```

marcos-vivar reviewed (approved):
LGTM! :shipit:
